File: lab/python/matematica_e_operadores/torre_hanoi_linear.py
def mover(origem, destino, nome_origem, nome_destino):
    # Se um dos pinos está vazio, só pode mover do outro
    if not origem:
        origem.append(destino.pop())
        print(f"Mova disco de {nome_destino} para {nome_origem}")
    elif not destino:
        destino.append(origem.pop())
        print(f"Mova disco de {nome_origem} para {nome_destino}")
    elif origem[-1] < destino[-1]:
        destino.append(origem.pop())
        print(f"Mova disco de {nome_origem} para {nome_destino}")
    else:
        origem.append(destino.pop())
        print(f"Mova disco de {nome_destino} para {nome_origem}")


def torre_hanoi_iterativa(n):
    # discos em ordem decrescente (maior embaixo)
    origem = list(range(n, 0, -1))
    destino = []
    auxiliar = []

    nomes = ("A", "C", "B")  # origem=A, destino=C, auxiliar=B

    total_movimentos = 2**n - 1

    # Se n for par, troca destino e auxiliar
    if n % 2 == 0:
        nomes = ("A", "B", "C")

    for i in range(1, total_movimentos + 1):
        if i % 3 == 1:
            mover(origem, destino, nomes[0], nomes[1])
        elif i % 3 == 2:
            mover(origem, auxiliar, nomes[0], nomes[2])
        else:
            mover(auxiliar, destino, nomes[2], nomes[1])


# Exemplo com 3 discos
torre_hanoi_iterativa(3)

File: lab/javascript/matematica_e_operadores/torre_hanoi_linear.js
function mover(origem, destino, nomeOrigem, nomeDestino) {
    if (origem.length === 0) {
        origem.push(destino.pop());
        console.log(`Mova disco de ${nomeDestino} para ${nomeOrigem}`);
    } else if (destino.length === 0) {
        destino.push(origem.pop());
        console.log(`Mova disco de ${nomeOrigem} para ${nomeDestino}`);
    } else if (origem[origem.length - 1] < destino[destino.length - 1]) {
        destino.push(origem.pop());
        console.log(`Mova disco de ${nomeOrigem} para ${nomeDestino}`);
    } else {
        origem.push(destino.pop());
        console.log(`Mova disco de ${nomeDestino} para ${nomeOrigem}`);
    }
}

function torreHanoiIterativa(n) {
    let origem = Array.from({ length: n }, (_, i) => n - i); // discos n..1
    let destino = [];
    let auxiliar = [];

    let nomes = ["A", "C", "B"];
    let totalMovimentos = Math.pow(2, n) - 1;

    if (n % 2 === 0) {
        nomes = ["A", "B", "C"];
    }

    for (let i = 1; i <= totalMovimentos; i++) {
        if (i % 3 === 1) {
            mover(origem, destino, nomes[0], nomes[1]);
        } else if (i % 3 === 2) {
            mover(origem, auxiliar, nomes[0], nomes[2]);
        } else {
            mover(auxiliar, destino, nomes[2], nomes[1]);
        }
    }
}

// Exemplo
torreHanoiIterativa(3);
